players/g1_player.py

```diff
@@ -29,12 +31,80 @@ def generate_choices(top_layer: np.ndarray, curr_level: np.ndarray) -> List[Choi
     return res
 
 
+def choose_next_player(now_turn: int, possible_players: List[int], served_situation: List[Dict[int, int]],
+                       top_layer: np.ndarray,
+                       curr_level: np.ndarray, rng: np.random.Generator, group_id: int) -> int:
+    if len(possible_players) == 1:
+        return possible_players[0]
+
+    # if now_turn = 0, it means for the first turn, we just randomly choose a person to pass to
+    if now_turn == 0 and len(possible_players) != 0:
+        return rng.choice(possible_players)
+
+    # if there are no candidates, which means our team is the last one to serve
+    if len(possible_players) == 0:
+        possible_players = [i for i in range(10) if i != group_id]
+
+    choices = generate_choices(top_layer, curr_level)
+    # if all ice cream has been taken
+    if len(choices) == 0:
+        return rng.choice(list(possible_players))
+
+    greedy_flavor_preference = defaultdict(list)
+    for player_index in possible_players:
+        player_served_situation = sorted(served_situation[player_index].items(), key=lambda item: -item[1])
+        for flavor, _ in player_served_situation:
+            greedy_flavor_preference[player_index].append(flavor)
+
+    max_score = -math.inf
+    next_player_list = set()
+    for player_index, flavor_preference in greedy_flavor_preference.items():
+        score_list = []
+        for choice in choices:
+            score_list.append((score(choice, flavor_preference), len(choice.flavors)))
+
+        score_list.sort(key=lambda x: -x[0])
+        remain = 24
+        player_max_score = 0
+        for score_num, count in score_list:
+            if remain < 0:
+                break
+            remain -= count
+            player_max_score += score_num
+
+        if max_score < player_max_score:
+            max_score = player_max_score
+            next_player_list = set()
+            next_player_list.add(player_index)
+        elif max_score == player_max_score:
+            next_player_list.add(player_index)
+
+    # if len(next_player_list) == 1, just return,
+    # if the length is larger than 1, we may choose randomly choose a from the set
+    # or use a weighted function(still needs to think)
+    if len(next_player_list) == 1:
+        return list(next_player_list)[0]
+    return rng.choice(list(next_player_list))
+
+
+# copy from the below f function
+def score(choice: Choice, flavor_preference) -> float:
+    res = 0
+    for flavor in choice.flavors:
+        res -= flavor_preference.index(flavor)
+    res /= len(choice.flavors)
+    res += choice.max_depth * 0.2
+    res += 0.01 * len(choice.flavors)
+    return res
+
+
 class Player:
     def __init__(self, flavor_preference: List[int], rng: np.random.Generator, logger: logging.Logger) -> None:
         self.flavor_preference = flavor_preference
         self.rng = rng
         self.logger = logger
         self.state = [0]
+        self.group_id = 1 - 1
```

Comment:
?

Reply:
it is just our group_id, I just remind myself to minus 1